### PR TITLE
BNAS KD Loss

### PR DESCRIPTION
**Pull request by jpablomch**
### Changes

Enables KD loss in BootstrapNAS.

Using KD loss improves the training efficiency and produces better search results. Tested on MobileNetV3-ImageNet and BERT-SQuAD v1.1. 

### Related tickets

N/A

### Tests

test_knowledge_distillation_training_process.

### CODE REVIEW

**Comment by jpablomch** [comment]
@ljaljushkin Please have a look. Thanks! 
@yzheng124

**Comment by ljaljushkin** [comment]
Should we deal with elastic kernel problem, first?
The design of kd loss will depend on the fix in my understanding.

File: tests/torch/nas/test_ps_controller.py [hunk 1]
```diff
@@ -63,6 +70,52 @@ def update_train_bn_adapt_section(nncf_config, bn_adapt_section_is_called):
...
+    mse = torch.nn.MSELoss().cuda()
+    model.train()
+    kd_model.train()
+    output_storage = []
+    for _, (input_, __) in enumerate(mock_dataloader):
+        input_ = input_.to(get_model_device(model))
+        output = model(input_)
+        kd_output = kd_model(input_)
+        output_storage.append(output)
+        loss = mse(output, kd_output)
```
Comment (ljaljushkin) [inline tests/torch/nas/test_ps_controller.py:112]: `run_actual` and `run_reference` looks identical, except calculation of the loss
why not extract this to the functor:
```python
def run_train(training_ctrl, model, kd_model, mock_dataloader, calc_loss_fn):
   ...
   loss = calc_loss_fn(output)
```

```python
def calc_loss_actual(output, training_ctrl):
    return training_ctrl.loss()
```

```python
def calc_loss_reference(output, kd_model):
    kd_output = kd_model(input_)
    return mse(output, kd_output)
```

```python
run_train(...,calc_loss_fn=partial(calc_loss_actual, training_ctrl), ..)
run_train(...,calc_loss_fn=partial(calc_loss_reference, kd_model), ..)
```

Comment (jpablomch) [reply tests/torch/nas/test_ps_controller.py]: @ljaljushkin Please check the refactoring of the test. Thanks! @yzheng124

**Review by ljaljushkin — APPROVED** [review]
LGTM, please fix pylint and we will merge it

**Comment by jpablomch** [comment]
run build

**Comment by ljaljushkin** [comment]
run pytorch pre-commit tests

**Comment by ljaljushkin** [comment]
> run build

FYI: https://github.com/openvinotoolkit/nncf/wiki/How-to-trigger-pre-commit-jobs

File: nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py [hunk 2]
```diff
@@ -88,15 +91,27 @@ def _get_algo_specific_config_section(self) -> Dict:
...
             elasticity_ctrl,
             self._bn_adaptation,
             self._progressivity_of_elasticity,
             schedule_params,
             self._lr_schedule_config,
+            compression_loss_func,
         )
 
+    def _build_compression_loss_function(self, model: NNCFNetwork) -> "PTCompressionLoss":
+        compression_builder = create_compression_algorithm_builder(self._algo_config)
```
Comment (ljaljushkin) [inline nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py:106]: Could you please verify what will happen if there's no KD algorithm section in the config? Will it return `ZeroCompressionLoss`?

Comment (yzheng124) [reply nncf/experimental/torch/nas/bootstrapNAS/training/progressive_shrinking_builder.py:106]: Yes. When nncf config doesn't include KD algorithm section, this function will return `ZeroCompressionLoss`. When KD algorithm is included, then this function will return `KnowledgeDistillationLoss`.
